File: src/01-easy/happy-numbers/solution.clj
(ns Solution
  (:gen-class))
(defn dss [n]
  (loop [x n s 0]
    (if (zero? x) s
      (let [d (mod x 10)] (recur (quot x 10) (+ s (* d d)))))))
(defn happy? [x]
  (loop [x x seen #{}]
    (cond (= x 1) true
          (seen x) false
          :else (recur (dss x) (conj seen x)))))
(defn -main [& _]
  (let [n (Integer/parseInt (read-line))]
    (dotimes [_ n]
      (let [s (read-line)
            x (reduce + (map #(let [d (- (int %) 48)] (* d d)) s))]
        (println (str s (if (happy? x) " :)" " :(")))))))

File: src/01-easy/happy-numbers/solution.js
const n = parseInt(readline());
for (let i = 0; i < n; i++) {
    const s = readline();
    const seen = new Set();
    let x = [...s].reduce((a, c) => a + (c.charCodeAt(0) - 48) ** 2, 0);
    while (x !== 1 && !seen.has(x)) { seen.add(x); x = [...String(x)].reduce((a, c) => a + (c.charCodeAt(0) - 48) ** 2, 0); }
    console.log(`${s} ${x === 1 ? ':)' : ':('}`);
}
